psydac/api/fem.py

```diff
@@ -880,11 +880,11 @@ def __init__(self, expr, kernel_expr, domain_h, space, *, vector=None,
         discrete_space            = test_space
 
 
-        space_quad_order = [qo - 1 for qo in get_quad_order(test_space)]
-        quad_order       = [qo + 1 for qo in (quad_order or space_quad_order)]
+        space_nquads = [qo - 1 for qo in get_nquads(test_space)]
+        nquads       = [qo + 1 for qo in (nquads or space_nquads)]
 
         # this doesn't work right now otherwise. TODO: fix this and remove this assertion
-        assert np.array_equal(quad_order, get_quad_order(test_space))
+        assert np.array_equal(nquads, get_nquads(test_space))
```

Comment:
I don't remember introducing these lines of code, but the reasoning behind it is the following:
- if `nquads` is `None` then `space_nquads` degree will be `qo-1` which implies that nquads = `qo-1+1` = `qo`
- if `nquads` is not `None` we assume that the user gave us the degree and not the order so we add one

Reply:
Then there are two ways to continue from here:
   - Either rename the "`__init__`-variable" `nquads` to `quad_order`, as we expect the user to indeed pass the desired quadrature order and not the number of quadrature points.
   ```
   # in DiscreteLinearForm(BasicDiscrete)
   def __init__(self, ..., quad_order=None, ...):
   ```
   - Or, now assuming that the user provides the number of quadrature points instead of the quadrature degree, we write
   ```
   space_nquads = get_nquads(test_space)
   nquads = [qo for qo in (nquads or space_nquads)] # as nquads can still have None list entries
   ```

If nquads is None, we set nquads = get_nquads(test_space). Do we want this additional degree of freedom in letting the user choose `nquads` himself? Does it make sense? Or is `get_nquads(test_space)` always the best choice? (working with splines, probably choosing get_nquads(test_space) results in exact integration?)

While no `DiscreteLinearForm` object is created explicitely (but only implicitly using discretize(...)?), the additional assertion
`assert np.array_equal(nquads, get_nquads(test_space))`
implies that
   1. the user provided variable `nquads` is actually the quuadrature order &
   2. we don't actually want to user to have a choice.

What do you think @saidctb @yguclu ?

Reply:
The underlying reason for this PR is that the user should be able to choose the number of quadrature points `nquads` directly, rather than providing the number `quad_order` which sounds like an order of accuracy but really is not.

In fact `quad_order` was the spline degree for which the mass matrix is computed exactly *if there is no mapping*. This was really confusing.

If the user does not provide `nquads`, Psydac should compute its value based on the accuracy required for computing the mass matrix. The name `quad_order` should not appear in the library any more.

Reply:
Okay, now I understand! That means that the code should look like this:
```
   space_nquads = get_nquads(test_space)
   nquads = [qo for qo in (nquads or space_nquads)]
```
The problem is though, that all our tests were passing the assertion
`assert np.array_equal(nquads, get_nquads(test_space))`
with a different "definition" of `nquads` - i.e., at a lot of places in the code, the variable nquads does not have the value that it should have, but instead 1 less (spline degree rather than number of points) than it's supposed value.

If you agree with this, then with your help I can look for all of these instances and change the code by adding `+1`.

This should then hopefully be the last change to this PR before we can merge it :sweat_smile:

Reply:
The real question is: how many quadrature points are used in the generated code?

Often we calculate the integral of an expression which contains functions belonging to different spaces. Clearly we need to evaluate the basis functions of the different spaces ***on the same quadrature grid***, regardless of the spline degree of each space.